### PR TITLE
Add compatibility tests for Kagome (the C++ implementation of Polkadot RE)

### PR DESCRIPTION
**Pull request by Harrm**
**Description**
Tests for SCALE codec and Merkle-Patricia Tree implementations from kagome, conforming to the current CLI of polkadot-re-tests, are provided. Kagome  is added as a Git submodule and its dependencies are added through Hunter package manager for automatical download during CMake building process.

### CODE REVIEW

File: build_test.sh [hunk 1]
```diff
@@ -4,7 +4,15 @@ SCRIPT_DIR=$(dirname -- "$(readlink -f -- "$BASH_SOURCE")")
 cd ${SCRIPT_DIR}
 
 mkdir -p build
-cd build
+
+# build kagome
```
Comment (drskalman) [inline build_test.sh]: I think this should be moved to test/testers/cpp-tester/CMakeLists.txt

**Review by drskalman — CHANGES_REQUESTED** [review]

**Comment by drskalman** [comment]
So Google tests doesn't compile using GCC 9.1 due 'Implicit declaration of copy functions' (http://timepp.github.io/doc/cpp14/depr.impldec.html). I think it means because a class has implicitly overloaded "operator=" they don't get the default copy constructor implicitly anymore (because the class probably had to do some extra stuff when it copies its instances). 

I tried to make using 
```
make CXXFLAGS='Wno-error=depricated-copy'
```
to no avil. I constantly get multiple of such error:
```
.hunter/_Base/5d68bcc/b6a2f74/dc6a1c5/Build/GTest/Source/googlemock/include/gmock/gmock-matchers.h:432:67: error: implicitly-declared ‘testing::internal::StrEqualityMatcher<std::__cxx11::basic_string<char> >::StrEqualityMatcher(const testing::internal::StrEqualityMatcher<std::__cxx11::basic_string<char> >&)’ is deprecated [-Werror=deprecated-copy]
  432 |   explicit PolymorphicMatcher(const Impl& an_impl) : impl_(an_impl) {}
```
 and so am unable to test the result. You might need to tell Hunter to explicitly use Wno-error=depricated-copy for Google tests somehow.

**Comment by Harrm** [comment]
> So Google tests doesn't compile using GCC 9.1 due 'Implicit declaration of copy functions' (http://timepp.github.io/doc/cpp14/depr.impldec.html). I think it means because a class has implicitly overloaded "operator=" they don't get the default copy constructor implicitly anymore (because the class probably had to do some extra stuff when it copies its instances).
> 
> I tried to make using
> 
> ```
> make CXXFLAGS='Wno-error=depricated-copy'
> ```
> 
> to no avil. I constantly get multiple of such error:
> 
> ```
> .hunter/_Base/5d68bcc/b6a2f74/dc6a1c5/Build/GTest/Source/googlemock/include/gmock/gmock-matchers.h:432:67: error: implicitly-declared ‘testing::internal::StrEqualityMatcher<std::__cxx11::basic_string<char> >::StrEqualityMatcher(const testing::internal::StrEqualityMatcher<std::__cxx11::basic_string<char> >&)’ is deprecated [-Werror=deprecated-copy]
>   432 |   explicit PolymorphicMatcher(const Impl& an_impl) : impl_(an_impl) {}
> ```
> 
> and so am unable to test the result. You might need to tell Hunter to explicitly use Wno-error=depricated-copy for Google tests somehow.

I added this flag to Hunter for GTest and GMock, it should be alright now.

**Comment by drskalman** [comment]
I'm still getting the same old error:

```
~/.hunter/_Base/5d68bcc/b6a2f74/dc6a1c5/Build/GTest/Source/googlemock/include/gmock/gmock-matchers.h:3975:65:   required from here
~/_Base/5d68bcc/b6a2f74/dc6a1c5/Build/GTest/Source/googlemock/include/gmock/gmock-matchers.h:432:67: error: implicitly-declared ‘testing::internal::StrEqualityMatcher<std::__cxx11::basic_string<char> >::StrEqualityMatcher(const testing::internal::StrEqualityMatcher<std::__cxx11::basic_string<char> >&)’ is deprecated [-Werror=deprecated-copy]
```
- I can't find where you added the -Werror=deprecated-copy flag. Could you point me to that?
- Could you compile with g++ 9.1.0 and check if it passes for you?

**Comment by Harrm** [comment]
> I'm still getting the same old error:
> 
> ```
> ~/.hunter/_Base/5d68bcc/b6a2f74/dc6a1c5/Build/GTest/Source/googlemock/include/gmock/gmock-matchers.h:3975:65:   required from here
> ~/_Base/5d68bcc/b6a2f74/dc6a1c5/Build/GTest/Source/googlemock/include/gmock/gmock-matchers.h:432:67: error: implicitly-declared ‘testing::internal::StrEqualityMatcher<std::__cxx11::basic_string<char> >::StrEqualityMatcher(const testing::internal::StrEqualityMatcher<std::__cxx11::basic_string<char> >&)’ is deprecated [-Werror=deprecated-copy]
> ```
> 
>     * I can't find where you added the -Werror=deprecated-copy flag. Could you point me to that?
> 
>     * Could you compile with g++ 9.1.0 and check if it passes for you?

It's added here https://github.com/soramitsu/kagome/blob/master/cmake/Hunter/config.cmake
in kagome repo. Are you sure you updated kagome submodule? I use g++ 9.1 on my machine where I work on the project and after this fix I never experienced this error again. If you find out that your version of kagome is up to date and the error is still appearing, please write to me, I will continue investingating how to eliminate this error.

**Comment by drskalman** [comment]
Sorry that was the problem. So I updated submodules recursively and that error was resolved. :-)

Now I got this error: 

```
In file included from ~/polkadot-spec/implementations/cpp/kagome/core/common/logger.cpp:6:
~/polkadot-spec/implementations/cpp/kagome/core/common/logger.hpp:9:10: fatal error: spdlog/fmt/ostr.h: No such file or directory
    9 | #include <spdlog/fmt/ostr.h>
      |          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [core/common/CMakeFiles/logger.dir/build.make:63: core/common/CMakeFiles/logger.dir/logger.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:601: core/common/CMakeFiles/logger.dir/all] Error 2
```
I thought I'll ask here in case you know the straight answer beforing digging deeper.

**Comment by Harrm** [comment]
> Sorry that was the problem. So I updated submodules recursively and that error was resolved. :-)
> 
> Now I got this error:
> 
> ```
> In file included from ~/polkadot-spec/implementations/cpp/kagome/core/common/logger.cpp:6:
> ~/polkadot-spec/implementations/cpp/kagome/core/common/logger.hpp:9:10: fatal error: spdlog/fmt/ostr.h: No such file or directory
>     9 | #include <spdlog/fmt/ostr.h>
>       |          ^~~~~~~~~~~~~~~~~~~
> compilation terminated.
> make[2]: *** [core/common/CMakeFiles/logger.dir/build.make:63: core/common/CMakeFiles/logger.dir/logger.cpp.o] Error 1
> make[1]: *** [CMakeFiles/Makefile2:601: core/common/CMakeFiles/logger.dir/all] Error 2
> ```
> 
> I thought I'll ask here in case you know the straight answer beforing digging deeper.

I was trying to reproduce it and it appears that on my system this error arises when I do git submodule update --init without --recursive

**Comment by lamafab** [comment]
Closing in favor of #82

**Pull request closed without merging**